### ISSUE TITLE
Differentiability of arbitrary design variables

### ISSUE THREAD

**Issue by fedegigio90** [NanoComp/meep#1747]
Hello there,
I would really appreciate if I could optimize a design with respect to arbitrary input design variables, different from the materialGrid chessboard - for instance a source position and angle. Do you confirm me that the only differentiable input object is the MaterialGrid as of now? I imagine there is no fundamental reason why the sources can't change in the adjoint formulation, right?

Thanks

**Comment by smartalecH** [comment]
You can differentiate w.r.t. whatever you want.

Currently, only a material grid works out of the box. But you can very easily map an arbitrary function (e.g. levelset) to the material grid and just backprop the gradients.

Since the Yee grid is a grid, it made sense to differentiate with respect to a grid as well. Then the user can arbitrarily map as needed from there.

**Comment by smartalecH** [comment]
> for instance a source position and angle.

Sounds like you're designing a grating coupler.

These parameters are not explicit design parameters, but rather "hyperparameters" (e.g. things that define the optimization problem and architecture itself). There are some ways to optimize hyperparameters, but I don't see how you could derive an adjoint source for something like that.

**Comment by stevengj** [comment]
In principle, it's pretty straightforward to add analytic gradients with respect to other parameters, e.g. differentiating with respect to parameters of the source ("b") terms is discussed in my notes: https://math.mit.edu/~stevenj/18.336/adjoint.pdf

The difficult here is software engineering: for parameters that flow through our C++ code, e.g. source parameters, it seems that we we will have to manually implement differentiability one parameter at a time.

**Comment by smartalecH** [comment]
Ch 2.6 of my [dissertation](http://hdl.handle.net/1853/67338) talks about this, if you are still interested.
